compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt

```diff
@@ -319,35 +313,47 @@ fun ProgramNode.compileToKotlin(
     mainFunctionBuilder.addParameter("host", Host::class)
     mainFunctionBuilder.addParameter("runtime", Runtime::class)
 
-    // TODO - figure out right way to get unique function names here
-    val hostFunNameMap: Map<Host, String> = this.hosts.associateWith { it.name + "function" }
-
-    // create a function for each host to run
-    for (entry in hostFunNameMap) {
+    // generate code for each host
+    for (host in this.hosts) {
         val curGenerator = BackendCodeGenerator(
             this,
-            entry.key,
+            host,
             codeGenerators,
             protocolComposer
         )
 
-        objectBuilder.addFunction(
-            curGenerator.generateHostFunction(
-                entry.key,
-                entry.value,
-                mainBody
-            )
+        val classBuilder = TypeSpec.classBuilder(
+            host.name.replaceFirstChar { it.uppercase() }
+        ).primaryConstructor(
+            FunSpec.constructorBuilder()
+                .addParameter("runtime", Runtime::class)
+                .build()
+        ).addProperty(
+            PropertySpec.builder("runtime", Runtime::class)
+                .initializer("runtime")
+                .build()
         )
+
+        for (function in this.functions) {
+            classBuilder.addFunction(
+                curGenerator.generateFunction(
+                    host,
+                    function.name.value.name,
+                    function
+                )
+            ).build()
+        }
+        objectBuilder.addType(classBuilder.build())
     }
 
     // create switch statement in main method so program can be run on any host
     mainFunctionBuilder.beginControlFlow("when (host)")
-    for (entry in hostFunNameMap) {
+    for (host in this.hosts) {
         mainFunctionBuilder.addStatement(
-            "%N -> %N(%N)",
-            entry.key.name,
-            entry.value,
-            "runtime"
+            "%N -> %N(%L).main()",
+            host.name,
+            host.name.replaceFirstChar { it.uppercase() },
```

Comment:
I don't like that this logic is duplicated. KotlinPoet allows you to to pass a `TypeSpec` object to resolve a `%N` reference. So you should create a map in the previous step (mapping hosts to their `TypeSpec`) and use that here instead.